klippy/extras/bed_mesh.py

```diff
@@ -0,0 +1,617 @@
+    def get_position(self):
+        # Return last, non-transformed position
+        if self.z_mesh is None:
+            # No mesh calibrated, so send toolhead position
+            return self.toolhead.get_position()
+        else:
+            # return current position minus the current z-adjustment
+            x, y, z, e = self.toolhead.get_position()
+            return [x, y, z - self.z_adjust, e]
+    def move(self, newpos, speed):
+        gc_z = newpos[2] - self.gcode.base_position[2]
+        if self.z_mesh is None or gc_z >= self.adj_stop:
+            # No mesh calibrated, or mesh leveling phased out.
+            self.z_adjust = 0.
+            self.toolhead.move(newpos, speed)
+        else:
+            prev_pos = self.get_position()
+            self.splitter.build_move(prev_pos, newpos)
+            while not self.splitter.traverse_complete:
+                split_move = self.splitter.split()
+                if split_move:
+                    self.z_adjust = self.splitter.z_offset
+                    self.toolhead.move(split_move, speed)
+                else:
+                    self.gcode.respond_error(
+                        "Mesh Leveling: Error splitting move ")
+    cmd_BED_MESH_OUTPUT_help = "Retrieve interpolated grid of probed z-points"
+    def cmd_BED_MESH_OUTPUT(self, params):
+        if self.z_mesh is None:
+            self.gcode.respond_info("Bed has not been probed")
+        else:
+            self.calibrate.print_probed_positions(self.gcode.respond_info)
+            self.z_mesh.print_mesh(self.gcode.respond, self.horizontal_move_z)
+    cmd_BED_MESH_CLEAR_help = "Clear the Mesh so no z-adjusment is made"
+    def cmd_BED_MESH_CLEAR(self, params):
+        self.set_mesh(None)
+
+
+class BedMeshCalibrate:
+    ALGOS = ['lagrange', 'bicubic']
+    def __init__(self, config, bedmesh):
+        self.printer = config.get_printer()
+        self.bedmesh = bedmesh
+        self.probed_z_table = None
+        self.build_map = False
+        self.end_gcode = config.get('end_gcode', None)
+        self.probe_params = {}
+        points = self._generate_points(config)
+        self._init_probe_params(config, points)
+        self.probe_helper = probe.ProbePointsHelper(config, self, points)
+        # Automatic probe:z_virtual_endstop XY detection
+        detect_offset = config.getboolean('detect_z_offset', False)
+        self.z_offset = None
+        if not detect_offset and config.has_section('stepper_z'):
+            zconfig = config.getsection('stepper_z')
+            home_pos_dir = zconfig.getboolean('homing_positive_dir', False)
+            # probe endstop is only used if homing toward the bed
+            if not home_pos_dir:
+                self.z_offset = zconfig.getfloat('position_endstop', None)
+                logging.debug(
+                    "bed_mesh: Using Z position_endstop for z_offset: %.4f"
+                    % (self.z_offset))
+        # Register MESH_BED_LEVING command
```

Comment:
Stale comment?

Reply:
Oops, I changed the name of the module and forgot to change that comment.